### DISCUSSION TITLE
Odd Issues when trying to rerender a tree after focus regained event

### DISCUSSION THREAD

**Discussion by miversen33** [nvim-neo-tree/neo-tree.nvim#1075] [Internal Development]
The title isn't super clear here so to better explain what I am doing, here is a series of commands I am running
```vim
:Neotree remote
:q
:Neotree remote
```

This series of events throws the following error
```vim
E5108: Error executing lua ...en/.local/share/nvim/lazy/nui.nvim/lua/nui/tree/init.lua:261: Expected Lua number
stack traceback:
        [C]: in function 'nvim_win_get_cursor'
        ...en/.local/share/nvim/lazy/nui.nvim/lua/nui/tree/init.lua:261: in function 'get_node'
        ...miversen/git/netman.nvim/lua/netman/ui/neo-tree/init.lua:358: in function 'navigate'
        ...ersen/git/neo-tree.nvim/lua/neo-tree/sources/manager.lua:488: in function 'navigate'
        ...miversen/git/neo-tree.nvim/lua/neo-tree/command/init.lua:177: in function 'do_show_or_focus'
        ...miversen/git/neo-tree.nvim/lua/neo-tree/command/init.lua:135: in function 'execute'
        ...miversen/git/neo-tree.nvim/lua/neo-tree/command/init.lua:143: in function '_command'
```

And for completeness, the snippet of code causing this within my repo is
```lua
function M.navigate(state)
-- ... Other stuff that doesn't matter
    state.tree:get_node()
end
```

The `remote` source is Netman, though its not terribly important here. I have faced this issue several times in the past and never really gotten past it. I discussed with @cseickel a bit in relation to this (see #887 for background details).

I am finally getting back around to addressing this, but I don't really know _how_. When I talked before on #887, I was advised that I need to just recreate the tree each time. Which is fine, I have a way to serialize down my tree. However I _can't_ just recreate my tree as attempting to execute a `render.show_nodes` with the previous tree throws this fun error

```
Error setting nodes:  ...en/.local/share/nvim/lazy/nui.nvim/lua/nui/tree/init.lua:36: duplicate node iddocker:
//dev_env/home/
```

It seems as though the tree still has the previous nodes in it. And I can't remove said nodes as the window that `nui` thought it had associated with the tree doesn't exist anymore (which is where the above error is coming from).

So I suppose my question is, how are we _supposed_ to do this?

It's a bit tough for me to provide a repoducible configuration _outside_ of my setup as it is directly related to how I am trying to use Neo-tree. I don't actually know if this is a bug or if I am just doing something wrong.

For clarity, I have pushed up the bugged code to a branch on netman. The specific [navigate](https://github.com/miversen33/netman.nvim/blob/d6c8ffac397a7652f48ecfb4a6246ecbf5203e59/lua/netman/ui/neo-tree/init.lua#L343-L369) code is here, though I think this is really down to a "how am I expected to do X" vs "where is the bug in my code" thing.

**Comment by miversen33** [comment]
I took some time to dig into the provided `state.tree` on a navigate (after leaving the window) and it definitely does have all the previous child nodes in it still.

What makes this interesting is that if I don't go too deep in the nodes, I _can_ redraw the tree with no issue. This is likely a red herring but its curious none the less. I will continue digging into this and see if I can discover what is actually happening in this process. Also if this should be an issue vs a discussion, please let me know and I can convert this :)

**Comment by miversen33** [comment]
Thanks discussion board for being a great rubber ducky! Turns out I just write bad code :upside_down_face: Deep in the bowels of my serialization code, I was apparently secretly duplicating children.

**Reply by cseickel** [reply]
Glad to be of service!
